### ISSUE TITLE
if-let

### ISSUE THREAD

**Issue by atroche** [Engelberg/better-cond#6]
Hi @Engelberg,

Here's one example of my use of better-cond: https://gist.github.com/atroche/774e0d23e78d1c78713a89b4d09e6146

Would you consider adding an `if-let` that returns early if its value is not nil? Or do you think I could get similar benefits by just extracting the function calls I'm using in my lets into smaller, better-named functions?

Cheers!

**Comment by Engelberg** [comment]
Thanks for sharing.

I've thought about `if-let`.  As I understand it, you are suggesting:

`:if-let [var expr] return-right-away-with-this-value-if-var-is-truthy-otherwise-continue`

(you actually said "not nil", but to be similar to Clojure's if-let in order to avoid confusion, it would need to return on any truthy value)

as an alternative to the pattern of:

`:let [var expr], var return-right-away-with-this-value-if-var-is-truthy-otherwise-continue`

One main issue with an `:if-let` clause is that it would complicate the macro -- right now the code leverages the fact that everything in the cond comes in pairs of two elements, and this kind of clause would have three.  Is this surmountable?  Sure, but it just seemed like a complexity that wasn't worth incurring without good cause.

better-cond 1.0.0 did workaround this by including something analogous to if-let -- it treated a binding vector in the test position as an if-let.  So, essentially, if you wanted an if-let, you'd write it like this:

``` clojure
(cond
  [var expr] return-right-away-with-this-value-if-var-is-truthy-otherwise-continue
  ...)
```

where that [var expr] line is treated as an if-let.

I had this in my personal version of the library that I have been using for several years, but I honestly never used it once, so I removed it in better-cond 1.0.1 because I feared it would be more confusing than helpful.  (There are no other differences between 1.0.0 and 1.0.1, so if you want to experiment with this feature, feel free to use 1.0.0).

Another issue is that I've spoken with several people who had opposite notions of what `:if-let` should do.  Some people think it should bail with the result if falsey, others if truthy.  I think the semantic confusion comes over the fact that :when-let continues if true, so some people (reasonably) expect :if-let to continue if true.

In any case, I've come to appreciate the extra precision that comes from having the test on a separate line than the binding.  I've found that when I rely on things like `if-let` and `when-let` I often run into problems, because mentally, when I use those constructs I'm really thinking that I expect non-nil, but those constructs really only care about truthiness, so a value of `false` messes up my expectations, resulting in a potentially hard-to-track-down bug.  Or sometimes, I'm expecting an empty sequence to be treated as false, but it actually evaluates to true because the function returns () rather than nil for empty.

So I like being as specific in my test as possible:

``` clojure
(cond
  :let [var expr]
  (nil? var) ...       ; more precise version of :when-let
  :let [var expr]
  (not-empty var) ...   ; more precise version of :if-let when I'm expecting to bail if var is non-empty seq
  :let [var expr]
  (not (false? var)) ...     ; more precise version of :if-let if expr is a something-or-false predicate
  :let [var expr]  
  var ...   ; no different than an if-let would be, but extremely clear about our intention
```

Being more precise like this has saved me a lot of debugging time by allowing me to catch mistaken assumptions about the values more quickly.  Most of your examples that are binding `*-problems` appear to be an instance where you might benefit from an explicit test for non-emptiness.  Only `not-allowed` seems to be an actual truthy test.

Those are the reasons I'm not in a rush to add `:if-let` -- more complexity to the macro, possible confusion about how to interpret it, and I've come to like the extra clarity of explicitly stating the test you want.
